### PR TITLE
Provide intra-fit() ckpt_path access to enable additional finetuning modalities

### PR DESCRIPTION
**Pull request by speediedan**
## What does this PR do?

This PR includes a few minor changes that enable callbacks to restore checkpoints iteratively within the context of `Trainer.fit()`.

An example use case that depends on access to ckpt_path (`fit_ckpt_path`) within fit() is available in a notebook-based tutorial ([lightning-tutorials PR](https://github.com/PyTorchLightning/lightning-tutorials/pull/115)).

### Does your PR introduce any breaking changes? If yes, please list them.

None
 
## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Yes :)
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @speediedan,

Great work there ! I believe we can improve the callback and maybe move the examples to https://github.com/PyTorchLightning/lightning-tutorials instead.

What do you think?

Best,
T.C

File: docs/source/advanced/finetuning_scheduler.rst [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+        This boring model has four finetuning phases:
+
+    .. rst-class:: sbs-blk1
+
+    .. code-block:: yaml
+      :linenos:
+      :emphasize-lines: 10
+
+            0:
```
[the remark this answers is not in the document]
Comment (speediedan) [reply docs/source/advanced/finetuning_scheduler.rst:91]: Agreed! I had planned to add an epoch-driven option in a subsequent version since I found using earlystopping-driven depth transitions as the dominant use case but I can add in this version. Initially thinking I'll add an optional max-epoch constraint within the default earlystopping-driven transitions with a default of -1 (purely early-stopping driven)

File: pl_examples/basic_examples/config/fts/RteBoolqModule_ft_schedule_albert_base.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+0:
```
[the remark this answers is not in the document]
Comment (speediedan) [reply pl_examples/basic_examples/config/fts/RteBoolqModule_ft_schedule_albert_base.yaml:1]: got it! will move to lightning-tutorials

File: docs/source/advanced/finetuning_scheduler.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply docs/source/advanced/finetuning_scheduler.rst:91]: I didn't think about this. I think it could be possible to abstract the fine-tuning strategy. EpochFinetuning / EarlyStopping one.

[the remark this answers is not in the document]
Comment (speediedan) [reply docs/source/advanced/finetuning_scheduler.rst]: True, though I wonder if the utility of an abstract EpochFinetuning class might be limited though given that it would essentially serve as a pointer to a set of epochs. I think I have an approach that is working well and would love if you take a look to see what you think. 

I've just pushed a commit to this PR that allows finetuning phase transitions to now be driven by EarlyStopping criteria, exclusively user-specified epoch transitions or a seamless composition of the two (the default mode). 

I've also enabled parameters to be fully specified or selected via regex. Included is a fair amount of validation and testing to ensure EarlyStopping and epoch-driven transition criteria are sensibly composed in a variety of contexts.

image

Lmk what you think sir! Thanks!

File: pl_examples/basic_examples/config/fts/RteBoolqModule_ft_schedule_albert_base.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+- model.classifier.bias
+- model.classifier.weight
+1:
+- model.albert.pooler.bias
+- model.albert.pooler.weight
+2:
+- model.albert.encoder.albert_layer_groups.0.albert_layers.0.ffn_output.bias
+- model.albert.encoder.albert_layer_groups.0.albert_layers.0.ffn_output.weight
+3:
+- model.albert.encoder.albert_layer_groups.0.albert_layers.0.ffn.bias
```
[the remark this answers is not in the document]
Comment (speediedan) [reply pl_examples/basic_examples/config/fts/RteBoolqModule_ft_schedule_albert_base.yaml]: Done! I've enabled parameters to be fully specified or selected via regex:
image

**Comment by speediedan** [comment]
since the scope of this PR has changed so much, to maximize clarity I've opened a [replacement PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/11696)
